Network statistics enhancements

jukkar opened the pull request:
Allow collection of network statistics / network interface.

jukkar committed:
net: stats: Make statistics collection per network interface

Instead of one global statistics, collect statistics information
separately for each network interface. This per interface statistics
collection is optional but turned on by default. It can be turned
off if needed, in which case only global statistics are collected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
sample: net: stats: Example how to use net_mgmt for statistics

Simple application that queries network statistics periodically
via net_mgmt interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on samples/net/stats/README.rst:
```diff
@@ -0,0 +1,58 @@
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/stats
+   :board: <board to use>
```
Since only qemu_* boards are meaningful here, it would be good to mention that:

        :board: <qemu_board to use>

jukkar replied:
Actually we are not restricted to only qemu with this sample. There is the sentence above this that says "A good way to run this sample application is with QEMU..." but we could run this with many different boards.

dbkinder replied:
True, but since this whole example talks about using QEMU, wouldn't it be good to use one of the qemu_* boards here?

dbkinder commented on samples/net/stats/README.rst:
```diff
@@ -0,0 +1,58 @@
+   :compact:
+
+If everything is configured correctly, the application will periodically print
+current network statistics to console.
```
``to the console``

jukkar commented:
Updated the readme file.

tbursztyka commented on subsys/net/ip/net_stats.c:
```diff
@@ -261,65 +261,67 @@ static int net_stats_get(u32_t mgmt_request, struct net_if *iface,
 	case NET_REQUEST_STATS_CMD_GET_ALL:
 		len_chk = sizeof(struct net_stats);
-		src = &iface->stats;
+#if defined(CONFIG_NET_STATISTICS_PER_INTERFACE)
```
all this below should go into the per-iface patch

tbursztyka commented on include/net/net_stats.h:
```diff
@@ -321,22 +321,22 @@ enum net_request_stats_cmd {
 #define NET_REQUEST_STATS_GET_ALL				\
 	(_NET_STATS_BASE | NET_REQUEST_STATS_CMD_GET_ALL)
 
-//NET_MGMT_DEFINE_REQUEST_HANDLER(NET_REQUEST_STATS_GET_ALL);
```
Only these are part of this patch (which should go first in the commit order).

tbursztyka commented on samples/net/stats/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+set(KCONFIG_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/Kconfig)
```
I am not sure of the relevancy of that sample: the net shell module is already a good example on how to use the stats api.

tbursztyka replied:
Sure, but it means then 2 places where maintenance will be needed when changing net stats (if that happens). Also I am not sure that many users will need to use that API, it's quite specific.

jukkar committed:
net: stats: Fix the net_mgmt statistics collection

Make sure that we return proper network statistics data if
someone asks it via network management interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
